File: packages/raisin-client-py/raisindb/branches.py
"""Branch management operations."""

from typing import TYPE_CHECKING, Any, Dict, List, Optional

from .protocol import RequestType, RequestContext

if TYPE_CHECKING:
    from .database import Database


class Branches:
    """Branch management operations."""

    def __init__(self, database: "Database"):
        """
        Initialize Branches manager.

        Args:
            database: Parent Database instance
        """
        self.database = database

    async def create(
        self,
        name: str,
        from_revision: Optional[str] = None,
        from_branch: Optional[str] = None,
    ) -> Any:
        """
        Create a new branch.

        Args:
            name: Branch name
            from_revision: Optional revision to branch from
            from_branch: Optional branch to branch from

        Returns:
            Created branch
        """
        context = self.database.get_context()
        payload: Dict[str, Any] = {"name": name}
        if from_revision is not None:
            payload["from_revision"] = from_revision
        if from_branch is not None:
            payload["from_branch"] = from_branch
        return await self.database.client.request_tracker.send_request(
            RequestType.BRANCH_CREATE, context, payload
        )

    async def get(self, name: str) -> Any:
        """
        Get a branch by name.

        Args:
            name: Branch name

        Returns:
            Branch information
        """
        context = self.database.get_context()
        payload = {"name": name}
        return await self.database.client.request_tracker.send_request(
            RequestType.BRANCH_GET, context, payload
        )

    async def list(self) -> List[Any]:
        """
        List all branches.

        Returns:
            List of branches
        """
        context = self.database.get_context()
        result = await self.database.client.request_tracker.send_request(
            RequestType.BRANCH_LIST, context, {}
        )
        return result if isinstance(result, list) else []

    async def delete(self, name: str) -> Any:
        """
        Delete a branch.

        Args:
            name: Branch name

        Returns:
            Deletion result
        """
        context = self.database.get_context()
        payload = {"name": name}
        return await self.database.client.request_tracker.send_request(
            RequestType.BRANCH_DELETE, context, payload
        )

    async def get_head(self, name: str) -> Any:
        """
        Get the HEAD revision of a branch.

        Args:
            name: Branch name

        Returns:
            HEAD revision
        """
        context = self.database.get_context()
        payload = {"name": name}
        return await self.database.client.request_tracker.send_request(
            RequestType.BRANCH_GET_HEAD, context, payload
        )

    async def update_head(self, name: str, revision: str) -> Any:
        """
        Update the HEAD revision of a branch.

        Args:
            name: Branch name
            revision: New HEAD revision

        Returns:
            Update result
        """
        context = self.database.get_context()
        payload = {"name": name, "revision": revision}
        return await self.database.client.request_tracker.send_request(
            RequestType.BRANCH_UPDATE_HEAD, context, payload
        )

    async def merge(
        self,
        source_branch: str,
        target_branch: str,
        strategy: Optional[str] = None,
        message: Optional[str] = None,
    ) -> Any:
        """
        Merge a source branch into a target branch.

        Args:
            source_branch: Source branch name
            target_branch: Target branch name
            strategy: Optional merge strategy
            message: Optional merge message

        Returns:
            Merge result
        """
        context = self.database.get_context()
        payload: Dict[str, Any] = {
            "source_branch": source_branch,
            "target_branch": target_branch,
        }
        if strategy is not None:
            payload["strategy"] = strategy
        if message is not None:
            payload["message"] = message
        return await self.database.client.request_tracker.send_request(
            RequestType.BRANCH_MERGE, context, payload
        )

    async def compare(self, branch: str, base_branch: str) -> Any:
        """
        Compare two branches to calculate divergence.

        Args:
            branch: Branch to compare
            base_branch: Base branch to compare against

        Returns:
            Comparison result with divergence information
        """
        context = self.database.get_context()
        payload = {"branch": branch, "base_branch": base_branch}
        return await self.database.client.request_tracker.send_request(
            RequestType.BRANCH_COMPARE, context, payload
        )

File: packages/raisin-client-java/src/main/java/com/raisindb/client/operations/Branches.java
package com.raisindb.client.operations;

import com.raisindb.client.protocol.RequestContext;
import com.raisindb.client.protocol.RequestType;

import java.util.*;
import java.util.concurrent.CompletableFuture;

/**
 * Branch management operations.
 */
public class Branches {

    private final Database database;

    public Branches(Database database) {
        this.database = database;
    }

    /**
     * Create a new branch.
     *
     * @param name         Branch name
     * @param fromRevision Optional revision to branch from
     * @param fromBranch   Optional branch name to branch from
     * @return CompletableFuture with created branch
     */
    public CompletableFuture<Object> create(String name, String fromRevision, String fromBranch) {
        RequestContext context = database.getContext(null, null);
        Map<String, Object> payload = new HashMap<>();
        payload.put("name", name);
        if (fromRevision != null) {
            payload.put("from_revision", fromRevision);
        }
        if (fromBranch != null) {
            payload.put("from_branch", fromBranch);
        }

        return database.getClient().getRequestTracker()
                .sendRequest(RequestType.BRANCH_CREATE, context, payload);
    }

    /**
     * Create a new branch from HEAD.
     *
     * @param name Branch name
     * @return CompletableFuture with created branch
     */
    public CompletableFuture<Object> create(String name) {
        return create(name, null, null);
    }

    /**
     * Get a branch by name.
     *
     * @param name Branch name
     * @return CompletableFuture with branch
     */
    public CompletableFuture<Object> get(String name) {
        RequestContext context = database.getContext(null, null);
        Map<String, Object> payload = new HashMap<>();
        payload.put("name", name);

        return database.getClient().getRequestTracker()
                .sendRequest(RequestType.BRANCH_GET, context, payload);
    }

    /**
     * List all branches.
     *
     * @return CompletableFuture with list of branches
     */
    public CompletableFuture<List<Object>> list() {
        RequestContext context = database.getContext(null, null);
        Map<String, Object> payload = new HashMap<>();

        return database.getClient().getRequestTracker()
                .sendRequest(RequestType.BRANCH_LIST, context, payload)
                .thenApply(result -> {
                    if (result instanceof List) {
                        return (List<Object>) result;
                    }
                    return Collections.emptyList();
                });
    }

    /**
     * Delete a branch.
     *
     * @param name Branch name
     * @return CompletableFuture with result
     */
    public CompletableFuture<Object> delete(String name) {
        RequestContext context = database.getContext(null, null);
        Map<String, Object> payload = new HashMap<>();
        payload.put("name", name);

        return database.getClient().getRequestTracker()
                .sendRequest(RequestType.BRANCH_DELETE, context, payload);
    }

    /**
     * Get the HEAD revision of a branch.
     *
     * @param name Branch name
     * @return CompletableFuture with HEAD revision
     */
    public CompletableFuture<Object> getHead(String name) {
        RequestContext context = database.getContext(null, null);
        Map<String, Object> payload = new HashMap<>();
        payload.put("name", name);

        return database.getClient().getRequestTracker()
                .sendRequest(RequestType.BRANCH_GET_HEAD, context, payload);
    }

    /**
     * Update the HEAD revision of a branch.
     *
     * @param name     Branch name
     * @param revision New HEAD revision
     * @return CompletableFuture with result
     */
    public CompletableFuture<Object> updateHead(String name, String revision) {
        RequestContext context = database.getContext(null, null);
        Map<String, Object> payload = new HashMap<>();
        payload.put("name", name);
        payload.put("revision", revision);

        return database.getClient().getRequestTracker()
                .sendRequest(RequestType.BRANCH_UPDATE_HEAD, context, payload);
    }

    /**
     * Merge a source branch into a target branch.
     *
     * @param sourceBranch Source branch name
     * @param targetBranch Target branch name
     * @param strategy     Merge strategy (optional, e.g., "fast_forward" or "three_way")
     * @param message      Merge commit message (optional)
     * @return CompletableFuture with merge result
     */
    public CompletableFuture<Object> merge(String sourceBranch, String targetBranch, String strategy, String message) {
        RequestContext context = database.getContext(null, null);
        Map<String, Object> payload = new HashMap<>();
        payload.put("source_branch", sourceBranch);
        payload.put("target_branch", targetBranch);
        if (strategy != null) {
            payload.put("strategy", strategy);
        }
        if (message != null) {
            payload.put("message", message);
        }

        return database.getClient().getRequestTracker()
                .sendRequest(RequestType.BRANCH_MERGE, context, payload);
    }

    /**
     * Merge a source branch into a target branch with default strategy.
     *
     * @param sourceBranch Source branch name
     * @param targetBranch Target branch name
     * @return CompletableFuture with merge result
     */
    public CompletableFuture<Object> merge(String sourceBranch, String targetBranch) {
        return merge(sourceBranch, targetBranch, null, null);
    }

    /**
     * Compare two branches to calculate divergence.
     *
     * @param branch     Branch name
     * @param baseBranch Base branch name to compare against
     * @return CompletableFuture with divergence information
     */
    public CompletableFuture<Object> compare(String branch, String baseBranch) {
        RequestContext context = database.getContext(null, null);
        Map<String, Object> payload = new HashMap<>();
        payload.put("branch", branch);
        payload.put("base_branch", baseBranch);

        return database.getClient().getRequestTracker()
                .sendRequest(RequestType.BRANCH_COMPARE, context, payload);
    }
}
